Add vignette for odin.dust, dust and mcstate

johnlees opened the pull request:
This attempts to document the whole workflow, with some slightly more complex examples than previously included.

Questions:

- Anything that would benefit from being added?
- Any objections to documenting all three packages working together here?
- Is the use of the cache for the MCMC runs in the vignette correct?

Closes #22

johnlees commented:
Not passing on Ubuntu 20.04 (devel):
```
ERROR: dependencies ‘curl’, ‘httr’, ‘oai’ are not available for package ‘socialmixr’
```
but fine on release. A CRAN error?

johnlees commented:
html here for convenience:
[vignette.zip](https://github.com/mrc-ide/mcstate/files/5261275/vignette.zip)

richfitz commented on DESCRIPTION:
```diff
@@ -28,8 +28,17 @@ Suggests:
     mockery,
     mvtnorm,
     testthat,
-    withr
```
Can you please sort these alphabetically?

richfitz commented on vignettes/deaths.txt:
```diff
@@ -0,0 +1,36 @@
+## Definition of the time-step and output as "time"
```
Any reason that this, and other odin models, are not a .R file? If it's just the linting drama that's tuneable, though lintr's file-ignore system seems broken to me

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+library(mcstate)
+library(rmarkdown)
+```
+
````
Probably we need some basic introductory stuff here about what the problem is we're trying to solve

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+)
+```
+
+```{r setup}
````
I know opinions vary, but I'm in favour of not explicitly loading the libraries

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+
+The above equations can straightforwardly be written out using the odin DSL:
+
+```r
````
This duplicates your model file, which will lead to confusion. Can you adapt the approach in dust to read in and print to screen.  Look for the part that says

````
```{r pkg_dust_r, echo = FALSE, results = "asis"}
r_output(readLines(file.path(path, "R/dust.R")))
```
````

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+n_{SI} &\sim B(S, 1 - e^{-\beta \frac{I}{N} \cdot dt)} \\
+n_{IR} &\sim B(I, 1 - e^{-\gamma \cdot dt)}
+\end{align*}$$
+
```
For the *very first* example, I'd wondered about using one of the builtin examples. So we have already `dust::dust_example("sir")` which implements this model. The reasoning here is that it's less stuff to start getting started with

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+
+```{r}
+library(odin.dust)
+gen_sir <- odin_dust("sir.txt")
````
I'd prefer an explicit `odin.dust::odin_dust()` here

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+gen_sir <- odin_dust("sir.txt")
+```
+
+## Adding age structure to the model
````
This feels like it's in the wrong vignette to me

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+This can be given as a parameter to the model by adding the argument `data = list(m = transmission)` when using the `new()` method on the `gen_age` generator.
+
+## Saving a model into a package
```
This is definitely in the wrong vignette - this belongs in odin.dust

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+to help automate the compilation of the model.
+
+# Running the SIR model with [`dust`](https://mrc-ide.github.io/dust/)
```
IMO all the material above here can be dropped and/or moved to different vignettes. I think as much as possible this vignette should focus on the "I want to do inference" problem probably. We can have as many vignettes as needed and IME keeping them fairly focussed, while hard, is necessary to keep them clear

johnlees replied:
Moving everything above the mcstate section to odin.dust

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+[`new()`](https://mrc-ide.github.io/dust/reference/dust.html#method-new) takes the data needed to run the model i.e. a list with any parameters defined as `user` in the odin code above, the value of the initial step $t_0$, and the number of particles, each for now can simply be thought of as an independent stochastic realisation of the model, but in the next step will be used when inferring model ...
+
+Additional arguments include the number of threads to parallelise the particles over, and the seed for the random number generator. The seed must be an integer, and using the same seed will ensure reproducible results for all particles. For threads to take effect, `openmp` must be enabled in the compilation step, which you can check by running:
```
Detail like this probably goes later on in a section on peformance?

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+                         seed = 1L)
+```
+
+Check the initial state is ten particles wide, with four states deep ($t$, $S$, $I$, $R$):
````
```suggestion
The initial state is ten particles wide, four states deep ($t$, $S$, $I$, $R$):
```

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+sir_model$run(20)
+```
+
+To change the parameters, the model does not need to be recompiled. We simply use the generator object to make a new model. Let's do this and increase the infection rate, and the population size. We will also use a smaller time step to calculate multiple transitions per unit time:
````
I think that the compilation part here is a red-herring (not the detail the user will be expecting). You might write this as

```suggestion
We can change the parameters, say by increasing the infection rate and the population size, by reinitalising the model. We will also use a smaller time step to calculate multiple transitions per unit time:
```

**However** this goes against advice in the dust vignette to use `$reset()` if you want to change parameters, so you might do this instead with that function

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+matlines(time, t(x[2, , ]), col = "#cc0044", lty = 1)
+matlines(time, t(x[3, , ]), col = "#999966", lty = 1)
+legend("left", lwd = 1, col = c("#8c8cd9", "#cc0044", "#999966"), legend = c("S", "I", "R"), bty = "n")
+```
````
Consider something above like

```
cols <- c(S = "#8c8cd9", I = "#cc0044", R = "#999966")
```

so that you can do

```
matplot(..., col = cols[["S"]])
matlines(..., col = cols[["I"]])
legend(..., col = cols, legend = name(cols))
```

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+# Inferring parameters with [`mcstate`](https://mrc-ide.github.io/odin.dust/)
+
+Typically these models will be used in conjunction with observed data to infer the most likely parameter values, most likely trajectories up to the present (nowcast), and make forecasts using these parameters. The `mcstate` package implements particle filter and pMCMC methods to make all of this this possible from `dust` models.
```
"most likely" looks incorrect to me as we don't offer any ML methods 😛

richfitz replied:
Also pMCMC needs defining, possibly link to the wikipedia page or an article?

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+A comparison function is needed to define a likelihood, i.e. the probability of observing this trajectory, given the data. This is combined with a prior to generate a posterior distribution for each parameter, which in the SIR model gives the distribution of likely values for $\beta$ and $\gamma$. 
+
+This is straightforward for a deterministic model, but for a stochastic model looking at a single run can easily bias these estimates. `mcstate` therefore implements a particle filter, also known as sequential Monte Carlo (SMC), which runs multiple trajectories, and at each stage where data is available resamples them in proportion to their likelihood. This produces an unbiased estimate of the li ...
```
Does a single run "bias" these estimates?

johnlees replied:
Yes this is wrong, I've read a review briefly and understand this a bit better now

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+- A set of observations to fit the model to, generated using `particle_filter_data()`.
+- A model to fit, which must be a dust generator.
+- A comparison function, which is an R function which calculates the likelihood of the state given the data.
```
given the data _at a particular point in time_

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+Many other methods are available on the dust object to support inference, but typically you will use the `mcstate` package instead of calling these directly. However, if you wish to simulate state space models with set parameters, this can be done entirely using the commands above from `dust`.
+
+# Inferring parameters with [`mcstate`](https://mrc-ide.github.io/odin.dust/)
```
Actually *this* might be where I would start the vignette :)

johnlees replied:
Yes, starting here

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+
+To demonstrate this, we will attempt to infer the parameters $\beta$ and $\gamma$ using daily case counts. In reality these counts would be measured, but here we generate them from a single run of the model, and adding some constant noise proportional to the epidemic size:
+
+```{r}
````
My suggestion here is that you provide the data as a fait accompli. You can tell the reader that they were generated from the model (or not), but read them from a csv or hide the sausage factory here

johnlees replied:
Excellent idea, this is confusing

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+This can then be processed with `particle_filter_data()` as follows:
+
+```{r paged.print=TRUE, layout="l-body-outset"}
+library(mcstate)
````
This is the second inclusion of this library, but I'd prefer ::

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+  if (is.na(observed$cases)) {
+    return(NULL)
+  }
+  if (is.null(pars$compare$exp_noise)) {
```
This if / else is complicating things here and was needed I think in order to help with testing. If you want to demonstrate tuning the parameter I would unconditionally use your else branch, otherwise I'd unconditionally use your if branch

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+       legend = c("S", "I", "R"), bty = "n")
+```
+
+Typically you will not use the `run()` function of the particle filter directly, and instead use the functions below to infer parameters.
````
If the user was developing a new inference method, doing a sensitivity analysis, or manually doing a grid search they might?

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+```{r}
+beta <- pmcmc_parameter("beta", 0.2, min = 0)
+gamma <- pmcmc_parameter("gamma", 0.1, min = 0, 
+                         prior = function(p) {
````
it is typical to omit the curly braces here

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+
+Note that we run four separate chains in serial here. If you have $m$ nodes available, each with $n$ cores, the most efficient way to run this step is to set `nthreads = n` on the filters, and run $m$ independent chains, one on each node setting `n_chains = 1`. These independent results can then be collected using `pmcmc_combine()`:
+
+```{r pmcmc_serial, cache=TRUE}
````
The cache will only impact your local run, and even then with horribly unpredictable results. I'll rig something up to make this nicer once it settles down as CRAN will not allow a 30 min vignette. However, perhaps just not running this would be best. TBH I am a bit confused here

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+processed_chains <- pmcmc_thin(pmcmc_run, burnin = 500, thin = 2)
+processed_chains
+apply(processed_chains$pars, 2, mean)
+```
````
Can we plot something here? It would be nice to see a basic plot of the parameters, and a prediction

johnlees replied:
I've put in a plot function as above, but the density of trajectories may leave something to be desired. `coda` is used to plot mcmc results (as we removed all plotting functions), and prediction is dealt with in a section below

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+### Optimal use of parallelisation
+
+Note that we run four separate chains in serial here. If you have $m$ nodes available, each with $n$ cores, the most efficient way to run this step is to set `nthreads = n` on the filters, and run $m$ independent chains, one on each node setting `n_chains = 1`. These independent results can then be collected using `pmcmc_combine()`:
```
I'm confused here about introducing both the within-chain parallelism and the multiple chains together

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+    progress = TRUE,
+    n_chains = 1
+  )
+pmcmc_distributed_run <- pmcmc_combine(pmcmc1, pmcmc2, pmcmc3)
```
Why is this called "distributed" if it was run in series?

richfitz replied:
ah, ok the idea is that these might have been run across different computers? This feels too much a detail of our workflow for an introductory vignette and might fit better in a vignette on scaling. It also avoids the very real problem of correctly getting the RNG streams correct here.

richfitz replied:
With all the above, I'd suggest dropping this section for now

johnlees replied:
Moved to another vignette, we can decide whether to keep or remove separately. I think it's an important advantage, so worth having somewhere from the get go. Agree it's too involved for the intro though

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+
+Putting this all together, the age structured SIR model is as follows:
+
+```r
````
This is not really used in the above? It looks like it would fit better in the odin vignette to me

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+
+The set up of the particle filter and observation function makes it easy to fit to multiple datastreams within a Bayesian framework. Let us model a disease where a proportion of infected cases $p_{\mathrm{death}} = 0.05$ die rather than recover by adding the following lines to the odin code:
+
+```r
````
again, dump out the real code here

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+We can simulate some data to fit to, as before. We will also make the case data patchy, making it missing on some days:
+
+```{r layout="l-body-outset", paged.print=TRUE }
+# Set up model
````
again, use some precooked csv?

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+  }
+  
+  ll_pois <- function(obs, model, exp_noise) {
+    if (is.na(obs)) {
```
The is.na branch needs explanation

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+  if (is.na(observed$cases) && is.na(observed$deaths)) {
+    return(NULL)
+  }
+  if (is.null(pars$compare$exp_noise)) {
```
Simplify as above

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+    exp_noise <- pars$compare$exp_noise
+  }
+  
+  ll_pois <- function(obs, model, exp_noise) {
```
This function might be lifted out to be a free function for ease of explanation - looks like it does not need the captured scope?

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+paged_table(combined_data)
+```
+
+We then need to add this into our comparison function, and in doing so we will also demonstrate the use of the `index` parameter for the particle filter. The only model outputs needed from the comparison function are $S$ and $D$. For models with a very large state space returning only those compartments required by the comparison function can be important for efficiency. An index function can be  ...
````
This paragraph hides the motivation in the middle

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+and this can be used to run the `pmcmc()` method, as above.
+
+## Running predictions
```
Save the reader bouncing around and put it with the simple model above?

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+                    seed = processed_chains$predict$seed)
+
+# Sample 10 particles
+mini_forecast <- forecast$state[,sample.int(dim(forecast$state)[2], size = 10),]
```
dim(x)[2] is nrow(x)

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+
+# Sample 10 particles
+mini_forecast <- forecast$state[,sample.int(dim(forecast$state)[2], size = 10),]
+time <- mini_forecast[1,1,-1]
```
Looks like linting is not on here by default. Can you look over the style and fix the bulk of these please? I can do this if you'd rather not

johnlees replied:
@richfitz linting the vignettes is a mystery to me unfortunately. If I run lintr::lint_package() they are ignored, same with lint_dir. You can run lint on the Rmd files directly and it works, but I don't know how to do this just from the .lintr file.

https://github.com/jimhester/lintr/blob/master/NEWS.md suggests this should run by default as of v2.0.1, so I'm not sure what's going on

johnlees replied:
Also I am unable to/incorrectly turning off linting of the odin code in CodeFactor

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,680 @@
+mini_forecast <- forecast$state[,sample.int(dim(forecast$state)[2], size = 10),]
+time <- mini_forecast[1,1,-1]
+
+matplot(time*dt, t(mini_forecast[2, , -1]), type = "l",
```
same colour treatment as above (and spaces around `*`)

richfitz commented on vignettes/sir_models_cache/html/__packages:
```diff
@@ -0,0 +1,5 @@
+odin.dust
+mcstate
+socialmixr
+rmarkdown
```
Please delete the cache files and add a suitable pattern to the gitignore

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,680 @@
+As before, save the file, and use `odin.dust` to compile the model:
+
+```{r}
+gen_age <- odin_dust("sirage.txt")
````
This appears never to be used? I'd remove this section entirely for now

richfitz reviewed (changes requested):
Sorry, lots of comments through here. I think something that is missing is a landing motivation. Something that says "you're a user and you have time series data that you want to fit a stochastic model to - htf are you going to do that?". The functional interface to mcstate with its index/compare/whatever functions probably needs a bit of high-level guidance too.

Docs are really hard.

I've attached the absolute skeleton of a doc I started working on ages ago. My hope had been to outsource most of the prose to someone else...

[mcstate.txt](https://github.com/mrc-ide/mcstate/files/5278055/mcstate.txt)

johnlees commented:
Will look at this more tomorrow, but for now:

- I've moved most of the start to https://github.com/mrc-ide/odin.dust/tree/vignette and will open a PR there when ready
- I've added more of an intro
- I've put the parallelisation information into its own vignette
- I've added some of your earlier vignette into it's own place, as I think it's useful to show that we tested the particle filter implementation correctly

I still need to:

- [ ] Turn on linting
- [ ] Fix the linting errors
- [ ] Check the parallelisation vignette
- [ ] Check the Kalman vignette

johnlees commented:
@richfitz linting the vignettes is a mystery to me unfortunately. If I run `lintr::lint_package()` they are ignored, same with `lint_dir`. You can run `lint` on the Rmd files directly and it works, but I don't know how to do this just from the `.lintr` file.

https://github.com/jimhester/lintr/blob/master/NEWS.md suggests this should run by default as of v2.0.1, so I'm not sure what's going on

johnlees commented:
Updated vignettes HTML: 
[mcstate_vignettes_v2.zip](https://github.com/mrc-ide/mcstate/files/5291589/mcstate_vignettes_v2.zip)

richfitz commented on .lintr:
```diff
@@ -4,4 +4,4 @@ linters: with_defaults(
     cyclocomp_linter = NULL
     )
-exclusions: list("tests/testthat.R")
+exclusions: list("tests/testthat.R", "vignettes/deaths.R")
```
lintr exclusions seem totally broken, but we can exclude in the codefactor interface

richfitz commented on vignettes/sir_incidence_data.txt:
```diff
@@ -0,0 +1 @@
+3,2,2,1,1,4,2,5,6,6,8,4,5,7,8,13,14,13,16,14,16,25,17,17,21,18,16,15,16,19,12,11,19,11,22,13,20,15,19,17,22,23,14,8,13,10,20,15,9,14,6,8,5,9,7,1,6,7,2,7,8,5,4,7,2,8,6,3,1,2,5,2,0,3,2,1,1,1,2,3,1,3,0,2,1,2,0,0,2,0,4,1,1,2,0,2,1,0,1,0
```
As discussed on teams:

* can this please be saved as a .csv with both columns
* can the script to generate it be put into the `scripts` directory; write as a standalone .R file that runs assuming that it is run from the package root. You can use `pkgload::load_all()` to load internal functions if needed

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,399 @@
+
+```{r}
+# Set up model
+incidence <- t(read.table("sir_incidence_data.txt", sep=",", header = F))
````
See the above comment about where this data should come from

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,399 @@
+dt <- 0.25
+sir_data <- mcstate::particle_filter_data(data = incidence,
+                                          time = "day",
+                                          rate = 1/dt)
```
This use of `dt` begs for more explanation

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,399 @@
+                                       seed = 1L)
+```
+
+We can now run the particle filter forward, which will run and resample 100 trajectories, and return the final likelihood. It is important to set the correct time-step $dt$ here, as we are using 0.25, rather than the default value of 1 defined in the odin model. We will also save the history, which allows us to plot the particle trajectories, as well as use the likelihood:
````
`$dt` should become `dt` in backticks for consistency

richfitz commented on vignettes/sir_models.Rmd:
````diff
@@ -0,0 +1,399 @@
+```{r fig.height=5, fig.width=7}
+forecast <- predict(processed_chains,
+                    steps = seq(400, 800, 4),
+                    n_threads = 4L,
````
throughout this vignette, please remove all uses of the `n_threads` argument.

The reasons for doing this are that we need to not use more than 2 threads on CRAN, but also it complicates writing the parallel vignette

richfitz commented on vignettes/sir_models.Rmd:
```diff
@@ -0,0 +1,399 @@
+# Set up model
+dt <- 0.25
+death_data <- read.table("death_data.csv", header = TRUE, sep = ",")
+true_history_deaths <- readRDS("deaths_true_history.rds")
```
All these rds files need saving in the same script as the .csv - can you save them with `saveRDS(..., version = 2)` so that we use a version of rds that works with older R?

richfitz reviewed (changes requested):
Minor tweaks, then I'll return the favour with a parallel vignette

richfitz reviewed (approved):
As discussed, the parallel vignette will get reworked at some point, but it's good to have a placeholder at least
